internal/manifest/oci_index.go

```diff
@@ -93,19 +115,54 @@ func (index *OCI1IndexPublic) ChooseInstance(ctx *types.SystemContext) (digest.D
 				Variant:      d.Platform.Variant,
 			}
 			if platform.MatchesPlatform(imagePlatform, wantedPlatform) {
-				return d.Digest, nil
+				if matchedDigest == "" {
+					matchedDigest = d.Digest
+				}
+				digest := instanceMatchesCompression(d, compressionAnnotation)
+				if digest != "" {
+					return digest, nil
+				}
+				continue
 			}
 		}
 	}
-
-	for _, d := range index.Manifests {
+	if matchedDigest != "" {
+		return matchedDigest, nil
+	}
+	for _, d := range manifests {
 		if d.Platform == nil {
-			return d.Digest, nil
+			digest := instanceMatchesCompression(d, compressionAnnotation)
+			if digest != "" {
+				return digest, nil
+			}
+			continue
 		}
 	}
 	return "", fmt.Errorf("no image found in image index for architecture %s, variant %q, OS %s", wantedPlatforms[0].Architecture, wantedPlatforms[0].Variant, wantedPlatforms[0].OS)
 }
 
+// ChooseInstanceByCompression parses blob as an oci v1 manifest index, and returns the digest
+// of the image which is appropriate for the current environment but if multiple digests matches
+// for a given platform it gives priority to the one which is configured in policy, as of now
+// only two options are allowed either prefer gzip or zstd.
+func (index *OCI1IndexPublic) chooseInstance(ctx *types.SystemContext, preferGzip types.OptionalBool) (digest.Digest, error) {
+	compressionAnnotation := ""
+	if preferGzip != types.OptionalBoolTrue {
+		compressionAnnotation = OCI1InstanceAnnotationCompressionZSTD
+	}
+	return instanceMatchesPlatformAndCompression(ctx, index.Manifests, compressionAnnotation)
+}
+
+func (index *OCI1Index) ChooseInstanceByCompression(ctx *types.SystemContext, preferGzip types.OptionalBool) (digest.Digest, error) {
+	return index.chooseInstance(ctx, preferGzip)
+}
+
+// ChooseInstance parses blob as an oci v1 manifest index, and returns the digest
+// of the image which is appropriate for the current environment.
+func (index *OCI1IndexPublic) ChooseInstance(ctx *types.SystemContext) (digest.Digest, error) {
+	return index.chooseInstance(ctx, types.NewOptionalBool(false))
```

Comment:
Everywhere: instead of `types.NewOptionalBool(constant)`, we can just use `types.OptionalBoolConstant`.